compiler/tests/evm_contracts.rs

```diff
@@ -290,65 +322,72 @@ fn test_assert() {
     case("return_u128_cast.fe", vec![], Some(uint_token(42))),
+    case("return_i128_cast.fe", vec![], Some(int_token(-3))),
     // binary operators
-    case("return_addition_u256.fe", vec![42, 42], Some(uint_token(84))),
-    case("return_addition_u128.fe", vec![42, 42], Some(uint_token(84))),
-    case("return_subtraction_u256.fe", vec![42, 42], Some(uint_token(0))),
-    case("return_multiplication_u256.fe", vec![42, 42], Some(uint_token(1764))),
-    case("return_division_u256.fe", vec![42, 42], Some(uint_token(1))),
-    case("return_pow_u256.fe", vec![2, 0], Some(uint_token(1))),
-    case("return_pow_u256.fe", vec![2, 4], Some(uint_token(16))),
-    case("return_mod_u256.fe", vec![5, 0], Some(uint_token(0))),
-    case("return_mod_u256.fe", vec![5, 2], Some(uint_token(1))),
-    case("return_mod_u256.fe", vec![5, 3], Some(uint_token(2))),
-    case("return_mod_u256.fe", vec![5, 5], Some(uint_token(0))),
-    case("return_bitwiseand_u256.fe", vec![12, 25], Some(uint_token(8))),
-    case("return_bitwiseand_u128.fe", vec![12, 25], Some(uint_token(8))),
-    case("return_bitwiseor_u256.fe", vec![12, 25], Some(uint_token(29))),
-    case("return_bitwisexor_u256.fe", vec![12, 25], Some(uint_token(21))),
-    case("return_bitwiseshl_u256.fe", vec![212, 0], Some(uint_token(212))),
-    case("return_bitwiseshl_u256.fe", vec![212, 1], Some(uint_token(424))),
-    case("return_bitwiseshr_u256.fe", vec![212, 0], Some(uint_token(212))),
-    case("return_bitwiseshr_u256.fe", vec![212, 1], Some(uint_token(106))),
+    case("return_addition_u256.fe", vec![uint_token(42), uint_token(42)], Some(uint_token(84))),
+    case("return_addition_i256.fe", vec![int_token(-42), int_token(-42)], Some(int_token(-84))),
+    case("return_addition_i256.fe", vec![int_token(-42), int_token(42)], Some(int_token(0))),
+    case("return_addition_u128.fe", vec![uint_token(42), uint_token(42)], Some(uint_token(84))),
+    case("return_subtraction_u256.fe", vec![uint_token(42), uint_token(42)], Some(uint_token(0))),
+    case("return_subtraction_i256.fe", vec![int_token(-42), int_token(-42)], Some(int_token(0))),
+    case("return_subtraction_i256.fe", vec![int_token(-42), int_token(42)], Some(int_token(-84))),
+    case("return_multiplication_u256.fe", vec![uint_token(42), uint_token(42)], Some(uint_token(1764))),
+    case("return_multiplication_i256.fe", vec![int_token(-42), int_token(-42)], Some(int_token(1764))),
+    case("return_multiplication_i256.fe", vec![int_token(-42), int_token(42)], Some(int_token(-1764))),
+    case("return_division_u256.fe", vec![uint_token(42), uint_token(42)], Some(uint_token(1))),
+    case("return_division_i256.fe", vec![int_token(-42), int_token(-42)], Some(int_token(1))),
+    case("return_division_i256.fe", vec![int_token(-1), int_token(1)], Some(int_token(-1))),
+    // TODO: Investigate what's wrong with this test
+    // case("return_division_i256.fe", vec![int_token(-42), int_token(42)], Some(int_token(-1))),
```

Comment:
Surprisingly, this test fails and I'm not entirely sure why. As we can see `-1 / 1 = -1` works but doing the same with any other number gives me surprises. I wonder if this is somehow related to https://github.com/ethereum/fe/pull/146/files#r538425906

Reply:
I think we'll need to write custom runtime functions for multiplication and division of signed integers. As I understand it, addition and subtraction works the same for signed and unsigned integers, but multiplication and division are different. `n / 1 = n` is the only case where the operations are the same.

It might be best to mark these operations as unimplemented and take care of them later.

Reply:
>but multiplication and division are different

Ah ok. I guess I'll take a solidity example and generate the YUL code from it to see what it compiles to.  

For the record, these multiplication cases work fine.
```
  case("return_multiplication_i256.fe", vec![int_token(-42), int_token(-42)], Some(int_token(1764))),
  case("return_multiplication_i256.fe", vec![int_token(-42), int_token(42)], Some(int_token(-1764))),
```

Reply:
> For the record, these multiplication cases work fine.

Interesting.

> Ah ok. I guess I'll take a solidity example and generate the YUL code from it to see what it compiles to.

Yeah, I'd like to see what Solidity is doing.

Reply:
For the record:

Multiplication
```solidity
      function checked_mul_t_int256(x, y) -> product {

                // overflow, if x > 0, y > 0 and x > (maxValue / y)
                if and(and(sgt(x, 0), sgt(y, 0)), gt(x, div(0x7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff, y))) { revert(0, 0) }
                // underflow, if x > 0, y < 0 and y < (minValue / x)
                if and(and(sgt(x, 0), slt(y, 0)), slt(y, sdiv(0x8000000000000000000000000000000000000000000000000000000000000000, x))) { revert(0, 0) }
                // underflow, if x < 0, y > 0 and x < (minValue / y)
                if and(and(slt(x, 0), sgt(y, 0)), slt(x, sdiv(0x8000000000000000000000000000000000000000000000000000000000000000, y))) { revert(0, 0) }
                // overflow, if x < 0, y < 0 and x < (maxValue / y)
                if and(and(slt(x, 0), slt(y, 0)), slt(x, sdiv(0x7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff, y))) { revert(0, 0) }

                product := mul(x, y)
            }
```

Division:

```solidity
           function checked_div_t_int256(x, y) -> r {
                if iszero(y) { revert(0, 0) }

                // overflow for minVal / -1
                if and(
                    eq(x, 0x8000000000000000000000000000000000000000000000000000000000000000),
                    eq(y, sub(0, 1))
                ) { revert(0, 0) }

                r := sdiv(x, y)
            }
```

Reply:
And tests are passing now with the `sdiv`. Still need to do some final cleanups after lunch before I'll ping you for another round of review :)